### PR TITLE
Osx port / some compilation adaptations

### PR DESCRIPTION
**Pull request by pthom**
Hi

Having some spare time this morning, I tried to build your project under osx (on a mac book pro).
I had to do some modifications in order to get it to compile under AppleClang : adding missing includes, adding openGL framework, etc. Nothing too fancy although. 

Now everything compiles successfully under OSX, and the unit tests pass.

However, I cannot run the SampleLauncher under OSX, as it fails inside `external/glfw/src/nsgl_context.m:137`:
> "NSGL: The targeted version of macOS does not support OpenGL 3.0 or 3.1 but may support 3.2 and above".

 If I disable the test altogether inside glfw, then I get an empty window. I suppose this might be due to the sad fact that apple is discontinuing its support for opengl. It is probably solvable although.

However feel free to look at the changes I did: I think they correct valid issues in the code.

Side note: I also compiled BabylonCpp under windows on my mac (using Parallels desktop virtual machine), but when I run SampleLauncher.exe , it triggers a bug inside Parallels desktop itself (and the machine is hung).

### CODE REVIEW

**Comment by samdauwe** [comment]
Hi pthom,

Thank you very much for trying out BabylonCpp and resolving the compiler errors on OSX. I don't own a Mac but I could add this build environment to the Travis CI configuration. At some point I also had a config for the Clang compiler in that configuration file, I will have a look if I can add it back.

Regarding that OpenGL 3.0 or 3.1 error. I forced these versions in the GLFW setup otherwise ImGui does not render anything. I am using the OpenGL 2 example and did not try yet with the OpenGL 3 example. I am using ImGui for the inspector to help with debugging the scenes.

Could you try the following test:
In the file _BabylonCpp/src/SampleLauncher/src/samples/sample_launcher.cpp_ replace the lines 511-515 by these:

```c++
glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
glfwWindowHint(GLFW_CONTEXT_CREATION_API, GLFW_NATIVE_CONTEXT_API);
glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
```
With these glfw hints you will be using OpenGL 3.3 and avoiding that error. Apple discontinuing OpenGL does not sound good, as OpenGL is still used a lot...

It also might be interesting to start the SampleLauncher with debug messaging enabled, that could reveal some issues during shader compilation. You can enable debug messages with the verbose option, like for instance:

```
./SampleLauncher --sample=CuboidScene --verbose
```

I did not try Parallels desktop virtual machine so I cannot comment on this, but if it makes your machine hang that it might be difficult to figure out how to resolve the issue. You might have better luck with testing in VirtualBox with a Linux OS, but there are also issues such as high CPU load (OpenGL is emulated I assume) and most of the scenes are not rendered properly (for instance those that use textures).

Hope it helps!
Sam

**Comment by pthom** [comment]
Hi,

> Could you try the following test:
In the file BabylonCpp/src/SampleLauncher/src/samples/sample_launcher.cpp replace the lines 511-515 by these:
> ...

I did this and it avoids the error inside glfw. Thanks!

However when I launch SampleLauncher I get an error during the shader compilation ("version 300 is  not supported")
````
./SampleLauncher --sample=CuboidScene --verbose

[2019-04-24 23:50:48.238] [ INFO] [GLRenderingContext::initialize] | Using GL version: 4.1 ATI-2.4.10
[2019-04-24 23:50:48.239] [ INFO] [Engine::Engine] | BabylonCpp engine (v3.3.0) launched
[2019-04-24 23:50:48.654] [ERROR] [Engine::_compileRawShader] | ERROR: 0:1: '' :  version '300' is not supported
ERROR: 0:1: '' : syntax error: #version
ERROR: 0:2: '' :  #version required and missing.
ERROR: 0:316: ';' : syntax error: Uniform blocks may not have identifiers in GLSL 140

[2019-04-24 23:50:48.654] [ERROR] [Engine::_compileRawShader] | #version 300 es
#define WEBGL2 
#define SPECULARTERM
#define HEMILIGHT0
#define LIGHT0
#define NORMAL
#define NUM_MORPH_INFLUENCERS 0
#define BonesPerMesh 0
... 
````

Do you have an idea about this?

Thanks!

**Comment by pthom** [comment]
One note concerning the glsl version error. I can get SampleLauncher to run successfully by replacing all instances of `"#version 300 es"` by `"#version 330"` in the shader sources

**Comment by samdauwe** [comment]
Hi pthom,

That is good news that the error is gone with changing the glfw hints and you get it working by replacing the version in the shader sources. That version is set using the function __processShaderConversion_ in the file _BabylonCpp/src/BabylonCpp/src/materials/effect.cpp_

When I have more time this evening I will check if I can get ImGui example_glfw_opengl3 working with these settings then we can make these changes the default.

Hopefully most examples are working, some interesting ones are:
./SampleLauncher --sample=RollercoasterScene
./SampleLauncher --sample=ImportDummy3Scene
./SampleLauncher --sample=ShaderMaterialPlanetShadertoyScene
./SampleLauncher --sample=HeightMapScene
./SampleLauncher --sample=PBRMaterialScene
...

Thanks!

**Comment by pthom** [comment]
Well, I tried the examples you mentioned : all work, except two of them: ImportDummy3Scene and PBRMaterialScene where I get a segfault. 

When it fails, the call stack is similar for both cases. The code is a bit hairy although; so below are some screenshots of the debugger status (with variables content) : this might shed some light on the issue. This is run under osx, so that it is a 64 bits process.

I ran: `./SampleLauncher --sample=PBRMaterialScene`

Below is a screenshot of the status inside the debugger in the reverse call stack order:

image

image

image

image

image

image

**Comment by samdauwe** [comment]
Hi,

thanks for testing on your side, at least you already have a view of the current state of this BabylonCpp project and the amount of code.

Regarding those examples that are not working, indeed both are related to loading the 'environment.dds' file. I assume the file and other textures are in the assets/textures directory in your build directory because you could load the HeightMap scene. The HeightMap scene uses stb_image to read the file, for .dds files I have written my own binary file reader. I think this binary file reader is not working properly on osx.

In the 3rd screenshot with breakpoint in the function __GetHalfFloatAsFloatRGBAArrayBuffer_ I noticed that the _arrayBuffer_ size is 1, on my machine this size is 1048736 which corresponds to the file size:

environment_dds

The function that reads the .dds file is depicted in the screenshot below:

read_binary_file

and dataOffset and dataLength values in the _to_array_ function seem to match yours:

to_array_pbr_material_scene

I am also on a 64bit machine, the size of an Uin16_t (16-bit unsigned integer) should match yours I assume.

So, according to me the binary file reader returns a byte array having a size that does not match the file size or there is an issue related to transferring that byte buffer through std::variant.

Hopefully this helps pinpointing the root cause of the issue.

Thanks!
Sam

**Comment by pthom** [comment]
> thanks for testing on your side, at least you already have a view of the current state of this BabylonCpp project and the amount of code.

Indeed, there is quite a lot of work, and the project is ambitious. Congrats!

> In the 3rd screenshot with breakpoint in the function _GetHalfFloatAsFloatRGBAArrayBuffer I noticed that the arrayBuffer size is 1, on my machine this size is 1048736 which corresponds to the file size.

Well, the issue with `arrayBuffer ` was a false alarm: the debugger was giving a false information. I have the same size on my side.

However, I did some more investigation:

* first, I would suggest to change the signature of `to_array` such a below : add a reference to the buffer param. Is there a reason you did not pass this param by reference (since it is const anyhow)? 

````
template <typename C, typename T>
std::vector<C> to_array(const std::vector<T> & buffer, size_t byteOffset,
                        size_t byteLength)
````

If I do this, then the program will seem to work as expected: 
image


* However this is just _by pure luck_ ! There is a very real buffer overflow, but in this case it works just by chance.

See below, I manually "instrumented" the code in order to check for overflows:

````cpp
template <typename C, typename T>
std::vector<C> to_array(const std::vector<T> & buffer, size_t byteOffset,
                        size_t byteLength)
{
  // C = uint16_t
  // T = uint8_t
  std::vector<C> bytes(byteLength);
  auto bytesize = bytes.size(); (void)bytesize;
  auto sizeofC = sizeof(C); (void)sizeofC;
  auto sizeofT = sizeof(T); (void)sizeofT;

  auto dst_nb_bytes = sizeofC * byteLength; (void) dst_nb_bytes;
  auto src_nb_bytes_remaining_after_offset = buffer.size() * sizeofT - byteOffset; (void) src_nb_bytes_remaining_after_offset;
  auto nb_bytes_to_copy = bytes.size() * sizeofC; (void) nb_bytes_to_copy;
  if (nb_bytes_to_copy  > dst_nb_bytes)
    raise(SIGINT); // break if memcpy is about to overflow
  if (nb_bytes_to_copy  > src_nb_bytes_remaining_after_offset)
    raise(SIGINT); // break if memcpy is about to overflow
  std::memcpy(bytes.data(), buffer.data() + byteOffset,
              bytes.size() * sizeof(C));
  return bytes;
}
````

And I find lots of instances where it overflows, for example:

image


* As a matter of fact, I do not understand the intent of this function: 
create a vector<uint16_t> from a vector<uint8_t> by using memcpy (i.e binary copy).
According to my understanding, each uint16 will then group two uint8 values but this is very much dependent on the endianess.


May I suggest you to use a tool like valgrind in order to check for potential overflows? Valgrind will do the same type of checks without having to tediously instrument all the calls.

I cannot run it on my side because I cannot run this program under my linux VM : openGL seems to not work with this project on my linux VM.

**Comment by samdauwe** [comment]
> > thanks for testing on your side, at least you already have a view of the current state of this BabylonCpp project and the amount of code.
> 
> Indeed, there is quite a lot of work, and the project is ambitious. Congrats!

Thanks! There is still plenty of work to be done, but I am already very happy with the scenes this engine is able to render. My next big milestones for the next 2-3 months are implementation of the Inpector and having glTF support.

> > In the 3rd screenshot with breakpoint in the function _GetHalfFloatAsFloatRGBAArrayBuffer I noticed that the arrayBuffer size is 1, on my machine this size is 1048736 which corresponds to the file size.
> 
> Well, the issue with `arrayBuffer ` was a false alarm: the debugger was giving a false information. I have the same size on my side.

Yes you are right, if the size would be indeed 1(like I thought yesterday), then it would be impossible to parse the header and having those byteOffset values that are correct, my mistake.

> However, I did some more investigation:
> 
>     * first, I would suggest to change the signature of `to_array` such a below : add a reference to the buffer param. Is there a reason you did not pass this param by reference (since it is const anyhow)?

Oh, yes I forgot to pass by ref...

>     * However this is just _by pure luck_ ! There is a very real buffer overflow, but in this case it works just by chance.

You are right again, there seems to be always a buffer overflow when calling this function:

```c++
template <typename C, typename T>
std::vector<C> to_array(const std::vector<T>& buffer, size_t byteOffset)
{
...
}
```

> 
> May I suggest you to use a tool like valgrind in order to check for potential overflows? Valgrind will do the same type of checks without having to tediously instrument all the calls.
> 
> I cannot run it on my side because I cannot run this program under my linux VM : openGL seems to not work with this project on my linux VM.

Thanks for further instrumenting the code fragment!
I isolated the code into a different function, by reading the file into a byte array and calling that to_array function with the parameters that resulted into the crash you got on your end and valgrind reported the following memory issues:

valgrind

The next step i did was comparing with the expected behavior in Babylon.js:

babylonjs

For a byte offset of 524408 the result vector should have a length of 262164 instead of the 524368. Then I found out that I had forgotten to subtract the _byteOffset_ value in the to_array function, so I changed the function as follows:

```c++
template <typename C, typename T>
std::vector<C> to_array(const std::vector<T>& buffer, size_t byteOffset)
{
  return to_array<C>(buffer, byteOffset,
                     (buffer.size() * sizeof(T) - byteOffset) / sizeof(C));
}
```

Now, about the purpose of this function. Probably it will become more clear if you look at that screenshot with the JavaScript version. In the debugger view on the right you can see the _ArrayBuffer_ holding the _environment.dds_ file data. This variable has the binary data interpreted as Int8Array, In16Array, Int32Array and Uin8Array structures. Furthermore, there is the line:

```javascript
var srcData = new Uint16Array(arrayBuffer, dataOffset);
```
This _Uint16Array(arrayBuffer, dataOffset);_ construction is further explained on [this page](
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray)

So in short, I needed that _to_array_ function to mimic JavaScript TypedArrays and to be able to interpret a binary file (or slices of it) as 8-bit, 16-bit or 32-bit numbers. This is not something I would normally do in C++, but to keep as close to the JavaScript version I added that to_array function. For the moment I have no solution yet for the endianess complexity. Hopefully the purpose of this _to_array_ is more clear to you?

Thanks and have a nice weekend!
Sam

**Comment by samdauwe** [comment]
Hi,

Last week I updated the ImGui OpenGL version from 2.x to 3.x. I tried setting the GLFW hints the same as your settings but some scenes did not render as they should. Strangely, I did not get any shader compilation error on the version number like you had. I did not try replacing all instances of "#version 300 es" by "#version 330" in the shader sources. This should make a difference I assume, and I will try this later this week. For now, I am still using GL 3.0 on my side.

I am now some steps further with the Inspector implementation, for instance:
./SampleLauncher --sample=LensFlaresScene --inspector

Results in the following output on my side:
lensflares_inspector

The inspector runs in a separate GLFW window. This was the easiest to implement and in that way I can exclude it from the build on untested platforms. I also render the GUI at 10 fps instead compared to the 60 fps of the scene. Getting the layout of that scene graph right kept me busy for some time, having the colored icons and selectable tree items is not really straightforward in ImGui.

Since the release of Babylon.js 4.0 I am back to my old routine, meaning porting the code during Monday-Friday and bug fixing/improving feature support during my weekends.

Best regards,
Sam
